File: Códigos_Celis/Python/DescomposicionLU.py
import numpy as np

def DescomposicionLU(A):
    N = len(A)
    matrices = []
    matrices2 = []
    A_actual = A


    for c in range(N):
        # Crear L
        L = np.zeros([N, N])
        L_inversa = np.zeros([N, N])
        pivote_actual = A_actual[c, c]

        for i in range(N):
            L_inversa[i,i] = 1.0

        for i in range(N):
            # Crear la matriz diagonal
            L[i,i]=1.0

        for j in range(c+1,N):
            L[j,c] = -A_actual[j,c]/pivote_actual
            L_inversa[j,c] = A_actual[j,c]/pivote_actual

        matrices.append(L)
        matrices2.append(L_inversa)

        #Actualizacion de la matriz A
        A_actual = L @ A_actual # Será U

    L_total = np.zeros((N, N))
    for i in range(N):
        L_total[i,i] = 1.0

    for i in range(len(matrices2)):
        L_total = L_total @ matrices2[i]

    return matrices,A_actual,L_total

A = np.array([[4,-2,1],[20,-7,12],[-8,13,17]],float)
L,U,L_inv=DescomposicionLU(A)

for i,L in enumerate(L):
    print("L",i)
    print(L)

print("\n Matriz U:")
print(U)

print("\n Matriz L inv:")
print(L_inv)


A = L_inv @ U

print("\n",A)

File: Codigos_Celis/Rust/src/descomposicion_lu.rs
use ndarray::{array, Array2};

pub fn descomposicion_lu(A:Array2<f64>) -> (Vec<Array2<f64>>, Array2<f64>, Array2<f64>) {
    let n = A.nrows();
    let mut matrices: Vec<Array2<f64>> = Vec::new();
    let mut matrices2: Vec<Array2<f64>> = Vec::new();

    let mut A_actual = A.clone();

    for c in 0..n{
        let mut L = Array2::<f64>::zeros((n,n));
        let mut L_inversa = Array2::<f64>::zeros((n,n));

        let pivote_actual = A_actual[[c,c]];

        for i in 0..n{
            L[[i,i]] = 1.0;
            L_inversa[[i,i]] = 1.0
        }

        for j in (c+1)..n{
            L[[j,c]] = -A_actual[[j,c]]/pivote_actual;
            L_inversa[[j,c]] = A_actual[[j,c]]/pivote_actual;
        }

        matrices.push(L.clone());
        matrices2.push(L_inversa.clone());

        A_actual = L.dot(&A_actual);
    }
    
    let mut L_total = Array2::<f64>::zeros((n,n));
    
    for i in 0..n{
        L_total[[i,i]] = 1.0;
    }
    
    for i in 0..matrices2.len(){
        L_total = L_total.dot(&matrices2[i]);
    }

    (matrices,A_actual,L_total)


}
